shared/clientstats/scrapers.go

```diff
@@ -0,0 +1,197 @@
+package clientstats
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prom2json"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+func scrapeProm(url string) (map[string]*dto.MetricFamily, error) {
+	mfChan := make(chan *dto.MetricFamily, 1024)
+	err := prom2json.FetchMetricFamilies(url, mfChan, http.DefaultTransport)
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]*dto.MetricFamily)
+	for fam := range mfChan {
+		ptr := fam
+		result[fam.GetName()] = ptr
+	}
+	return result, nil
+}
+
+type validatorScraper struct {
+	url string
+}
+
+func (vc *validatorScraper) Scrape() (io.Reader, error) {
+	log.Infof("Scraping validator at %s", vc.url)
+	pf, err := scrapeProm(vc.url)
+	if err != nil {
+		return nil, nil
+	}
+
+	cs, err := populateCommonStats(pf)
+	if err != nil {
+		return nil, err
+	}
+	vs, err := populateValidatorStats(pf)
+	if err != nil {
+		return nil, err
+	}
+	vs.CommonStats = cs
+
+	b, err := json.Marshal(vs)
+	return bytes.NewBuffer(b), err
+}
+
+func NewValidatorScraper(promExpoURL string) Scraper {
+	return &validatorScraper{
+		url: promExpoURL,
+	}
+}
+
+type beaconNodeScraper struct {
+	url string
+}
+
+func populateCommonStats(pf map[string]*dto.MetricFamily) (CommonStats, error) {
+	cs := CommonStats{}
+	cs.ClientName = "prysm"
+	var f *dto.MetricFamily
+	var m *dto.Metric
+	var ok bool
+
+	f, ok = pf["process_cpu_seconds_total"]
```

Comment:
If these hard-coded strings change in the future (although unlikely), it would be easier if they lived as consts at the top of the file perhaps

Reply:
IMO in an instance like this where the value is used in one place it is more readable to keep the strings inline, but don't feel super strongly about it.